feat(mcp-server): DLT-2787 implement 3-tool search architecture with smart filtering

belumontoya opened the pull request:
## Obligatory GIF (super important!)

  ![Search Magic](https://c.tenor.com/ACBGhLB0v0gAAAAC/tenor.gif)

  ## :hammer_and_wrench: Type Of Change

  - [x] Feature

  ## :book: Jira Ticket

  DLT-2787

  ## :book: Description

  Implements 3 search tools for the Dialtone MCP Server to help AI assistants find and recommend
   Dialtone utilities, tokens, and components.

  **What we built:**
  - `search_utility_classes` - Find CSS classes (d-p8, d-d-flex, d-w100p)
  - `search_tokens` - Find design tokens (--dt-color-foreground-primary, --dt-space-400)
  - `search_components` - Find Vue components (DtButton, DtModal, DtCheckbox)

  **How it works:**
  - Simple AND-logic search (all query words must match)
  - Px/rem OR matching (8px finds both "8px" and "0.8rem")
  - Smart filter removes deprecated items and swaps discouraged items with alternatives
  - Returns results in natural order (no sorting)

  **Test results:** 95% success rate (73/77 tests passing)

  ## :bulb: Context

  AI assistants need to search Dialtone to recommend the right classes, tokens, and components.
  The 3-tool architecture lets AI choose the right search based on user intent and prevents
  result flooding.

  ## :pencil: Checklist

  - [x] I have ensured no private Dialpad links or info are in the code or pull request
  description (Dialtone is a public repo!).
  - [x] I have reviewed my changes.
  - [x] I have added all relevant documentation.
  - [x] I have considered the performance impact of my change.

  ## :camera: Screenshots / GIFs

  N/A - Backend search functionality

  ## :link: Sources

  - Dialtone docs site search:
  `apps/dialtone-documentation/docs/.vuepress/baseComponents/tokens/AllTokens.vue`
  - Uses metadata from DLT-2798

  ---

  ## Test Prompts

  **Utility Classes:**
  How do I add 8px of padding on the right side?

  **Tokens:**
  What's the primary foreground color token?

  **Components:**
  I need a button component for my form

  **Smart Filter:**
  How do I make text bold?
  Should suggest composed typography classes, not d-fw-bold (discouraged)
  
  
<img width="699" height="716" alt="Screenshot 2025-10-29 at 6 00 13 PM" src="https://github.com/user-attachments/assets/1431321c-de77-4df2-bdea-14698aba7cca" />
<img width="705" height="820" alt="Screenshot 2025-10-29 at 6 04 16 PM" src="https://github.com/user-attachments/assets/e8250b53-22ae-409e-a856-fceec11dbcfe" />
<img width="705" height="513" alt="Screenshot 2025-10-29 at 6 04 54 PM" src="https://github.com/user-attachments/assets/7e238a1c-16e7-4f3b-a350-44d5a6d7759b" />
<img width="695" height="835" alt="Screenshot 2025-10-29 at 6 06 59 PM" src="https://github.com/user-attachments/assets/ba0eea34-4af8-420a-8fcc-d48e5e545fa7" />

juliodialpad reviewed (approved):
Great work here!

I'd prefer to have 1 regex to avoid repetitive code, here's a regex I was testing and might work.

`/-@su1\b|-var\(--su1\)|[^-]var\(--sun1\)/g`

But if you've tested and the way you made it works better I'm ok with that. Thanks!

braddialpad commented:
> Great work here!
> 
> I'd prefer to have 1 regex to avoid repetitive code, here's a regex I was testing and might work.
> 
> `/-@su1\b|-var\(--su1\)|[^-]var\(--sun1\)/g`
> 
> But if you've tested and the way you made it works better I'm ok with that. Thanks!

Nice, yeah that's better. Works for all the tests I tried. Replaced it.

belumontoya committed:
adding placeholder tool

belumontoya committed:
fixing schema without zod

belumontoya committed:
refining the search on css utilities

belumontoya committed:
adding token search

belumontoya committed:
search tool for tokens and components

belumontoya committed:
1. Unified result structure across all three search types:
  {
    type: 'utility-class' | 'design-token' | 'component',
    tier: 1 | 2 (or 3 for now),
    name: string,
    details: { ...type-specific data },
    metadata: {...} | null
  }
  2. Updated all search functions:
    - searchUtilityClasses() - Already had tiers, now uses unified format
    - searchTokens() - Added tier assignment (Tier 1 = all words match name, Tier 2 = any word
  or value match)
    - searchComponents() - Mapped priority to tiers (Tier 1 = exact match, Tier 2 = partial
  match)
  3. Updated all format functions:
    - formatResults() - Now uses result.name and result.details.properties
    - formatTokenResults() - Now uses result.name and result.details.allThemes
    - formatComponentResults() - Now uses result.name and result.details.

belumontoya committed:
Merge remote-tracking branch 'origin/staging' into DLT-2787-search-tool

belumontoya committed:
Breaking change: Replaced single search_dialtone tool with 3 separate tools
  for better AI tool selection and to prevent cross-type result flooding.

francisrupert commented:
Sweetness. Defer to everyone else.

ninarepetto reviewed (changes requested):
I like the way this is going.
The functions look good and the screenshots are amazing. 
The build is failing for me because the types are missing for `index.ts`. 
Also I think this branch is not up to date with staging because I am not getting the metadata in the resources.

belumontoya committed:
Merge branch 'staging' into DLT-2787-search-tool

belumontoya committed:
fix lintin issues

belumontoya commented:
> I like the way this is going. The functions look good and the screenshots are amazing. The build is failing for me because the types are missing for `index.ts`. Also I think this branch is not up to date with staging because I am not getting the metadata in the resources.

nina! you were right, I wasn't up to date, just pushed changes if you wanna take another look. I also added a JS file that can test queries directly to the search for you, i will remove it later. just run 
```node test-search.js```
you can also change the queries to be sure. thanks! 

cc @iropolo

ninarepetto commented:
This is perfection:

<img width="767" height="741" alt="image" src="https://github.com/user-attachments/assets/87d857af-295a-4720-8ab7-1051285ecb76" />

belumontoya committed:
fixing ts errors

ninarepetto reviewed (approved):
The build errors are fixed and it's working great! 🎉 
The `index.ts` is a bit long, but we can separate it later.

braddialpad reviewed (approved):
I would agree with Nina's comment that index.ts is overly complex and needs to be abstracted more, but I'm totally fine with merging this as is.

belumontoya commented:
> I would agree with Nina's comment that index.ts is overly complex and needs to be abstracted more, but I'm totally fine with merging this as is.

i can totally do that now, and we can forget about it

iropolo commented:
```
ussing tokens, how I can get the 73px of spacing?
```

```

I'll search for spacing tokens in Dialtone to find the appropriate token for 73px spacing.

Let me search for spacing tokens more broadly to see what values are available:

No Exact 73px Spacing Token Available
Dialtone does not have a token that equals exactly 73px. The spacing system uses a base of 8px with rem units.

Closest Options:
Looking at the spacing scale (assuming 1rem = 10px based on the system):

--dt-space-720 = 7.2rem = 72px (closest match, 1px off)
--dt-space-730 = 8.4rem = 84px (11px over)
Recommended Approaches:
Use the closest token (72px):
css
.your-element {
  margin: var(--dt-space-720);
}
Combine tokens if you need exactly 73px:
css
.your-element {
  margin: calc(var(--dt-space-720) + var(--dt-space-100)); /* 72px + 1px = 73px */
}
Reconsider the design - Dialtone's spacing system is based on 8px increments for consistency. Consider using 72px (--dt-space-720) or 80px (--dt-space-700 = 6.4rem = 64px) instead to align with the design system.
Note: The base spacing unit is --dt-space-base = 8px, and the system scales from there using rem units.
```

Seems like the answer is okay, but what about the last part? 

->

```
Reconsider the design - Dialtone's spacing system is based on 8px increments for consistency. 
Consider using 72px (--dt-space-720) or 80px (--dt-space-700 = 6.4rem = 64px) instead to align with the design system.

Note: The base spacing unit is --dt-space-base = 8px, and the system scales from there using rem units.
```

Could we somehow limit the additional comments? or it will always suggest something?

belumontoya committed:
less complexity in methods

belumontoya commented:
> ```
> ussing tokens, how I can get the 73px of spacing?
> ```
> 
> ```
> 
> I'll search for spacing tokens in Dialtone to find the appropriate token for 73px spacing.
> 
> Let me search for spacing tokens more broadly to see what values are available:
> 
> No Exact 73px Spacing Token Available
> Dialtone does not have a token that equals exactly 73px. The spacing system uses a base of 8px with rem units.
> 
> Closest Options:
> Looking at the spacing scale (assuming 1rem = 10px based on the system):
> 
> --dt-space-720 = 7.2rem = 72px (closest match, 1px off)
> --dt-space-730 = 8.4rem = 84px (11px over)
> Recommended Approaches:
> Use the closest token (72px):
> css
> .your-element {
>   margin: var(--dt-space-720);
> }
> Combine tokens if you need exactly 73px:
> css
> .your-element {
>   margin: calc(var(--dt-space-720) + var(--dt-space-100)); /* 72px + 1px = 73px */
> }
> Reconsider the design - Dialtone's spacing system is based on 8px increments for consistency. Consider using 72px (--dt-space-720) or 80px (--dt-space-700 = 6.4rem = 64px) instead to align with the design system.
> Note: The base spacing unit is --dt-space-base = 8px, and the system scales from there using rem units.
> ```
> 
> Seems like the answer is okay, but what about the last part?
> 
> ->
> 
> ```
> Reconsider the design - Dialtone's spacing system is based on 8px increments for consistency. 
> Consider using 72px (--dt-space-720) or 80px (--dt-space-700 = 6.4rem = 64px) instead to align with the design system.
> 
> Note: The base spacing unit is --dt-space-base = 8px, and the system scales from there using rem units.
> ```
> 
> Could we somehow limit the additional comments? or it will always suggest something?

im just sending the search result back, whatever the AI adds its on their side. I can influence a bit but that is all that is, influence. it can ignore my comments, or remove them. or follow them.

i also dont see it bad that will recommend using the right approach, you can ignore it but I dont think we want the AI teaching you how to hack the Design System

iropolo commented on packages/dialtone-mcp-server/README.md:
````diff
@@ -0,0 +1,176 @@
+```
+
+**Query Format:** Use keywords only, not questions.
+- ✓ Good: `padding 8px`, `color primary`, `button`
````
haha, funny that my tests were like how Nina did... doing questions 🤪

`Hey, could you give me the token for font size 15px please? tyvm 🫰"

edit:
I see this is the new interactive mode!

iropolo commented on packages/dialtone-mcp-server/README.md:
````diff
@@ -0,0 +1,176 @@
+# Navigate to the package directory first!
+cd packages/dialtone-mcp-server
+
+# Then run the interactive tool
+pnpm run interactive
+```
````
I will add my comments of interactive in this section.

iropolo replied:
Curious, it was intuitive to me to add " " in my query. 
But seems like it adds it beforehand. Then it doesn't work.

<img width="575" height="677" alt="image" src="https://github.com/user-attachments/assets/1a53efa0-90d5-4d30-97f6-dc4d7fcbd906" />

iropolo replied:
Searching for components, in this case Input, seems like it return also components that includes input?
hmm, tooltip doesn't has any input, just in a prop description.

<img width="575" height="677" alt="image" src="https://github.com/user-attachments/assets/7a1c0b21-a66a-4f3a-ac3b-b213a5050f09" />

iropolo replied:
huh yeah seems like the interactive mode does a more general finding.

<img width="572" height="262" alt="image" src="https://github.com/user-attachments/assets/a123a3ed-e7c9-48a3-a36b-daba28d36fe3" />

iropolo replied:
I guess its okay, I don't see many value to have it or using it over the claude/windsurf.

belumontoya replied:
your right 🤦‍♀️  this should test directly the search tools not a minimice version. that is the idea, to be able to test with queries the search tools without using AI, is not a replacement but if dialtone changes or re implements we can use this to test that the logic still translates and works without implementing it to any AI

iropolo reviewed (approved):
Looks great, I tested it using windsurf and it is very useful.

Thanks for the README file!

belumontoya committed:
improving the component search

belumontoya commented:
ok last time! @ninarepetto I just improve the component search, if you ask for a component it should match more easily now and be top of the list 🤞

ninarepetto reviewed (commented):
🚀
